plan/expression_rewriter.go

```diff
@@ -913,16 +913,23 @@ func (er *expressionRewriter) inToExpression(lLen int, not bool, tp *types.Field
 			return
 		}
 	}
-	leftArg := er.ctxStack[stkLen-lLen-1]
-	leftEt, leftIsNull := leftArg.GetType().EvalType(), leftArg.GetType().Tp == mysql.TypeNull
+	args := er.ctxStack[stkLen-lLen-1:]
+	leftEt, leftIsNull := args[0].GetType().EvalType(), args[0].GetType().Tp == mysql.TypeNull
 	if leftIsNull {
 		er.ctxStack = er.ctxStack[:stkLen-lLen-1]
 		er.ctxStack = append(er.ctxStack, expression.Null.Clone())
 		return
 	}
+	if leftEt == types.ETInt {
+		for i := 1; i < len(args); i++ {
+			if c, ok := args[i].(*expression.Constant); ok {
+				args[i] = expression.RefineConstantArg(er.ctx, c, opcode.EQ)
+			}
+		}
+	}
 	allSameType := true
-	for i := stkLen - lLen; i < stkLen; i++ {
-		if er.ctxStack[i].GetType().Tp != mysql.TypeNull && er.ctxStack[i].GetType().EvalType() != leftEt {
+	for _, arg := range args[1:] {
+		if arg.GetType().Tp != mysql.TypeNull && expression.GetAccurateCmpType(args[0], arg) != leftEt {
```

Comment:
Can we collect all the argument types and merge them to one type, then just cast all the expressions to this type and always use `in`?

Reply:
How about build DNF with different types?
`(intValLeft in (intA, intB, intC)) or (floatValLeft in (floatA, floatB))`?

Reply:
This is okay.

Reply:
But currently this is not very necessary to do this. This case can't push down to tikv, and can't be used to calculate range.